src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -259,15 +276,58 @@ impl ScalarFunc {
         }
         self.children[i as usize].eval_string(ctx, row)
     }
+
+    #[inline]
+    pub fn trim_1_arg<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        trim(&s, " ", TrimDirection::Both)
+    }
+
+    #[inline]
+    pub fn trim_2_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pat = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
+        trim(&s, &pat, TrimDirection::Both)
+    }
+
+    #[inline]
+    pub fn trim_3_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pat = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
+        let direction = TrimDirection::from_i64(try_opt!(self.children[2].eval_int(ctx, row)));
+        trim(&s, &pat, direction.unwrap())
```

Comment:
Hi, the direction value is passed from tidb, so tidb is responsible to parse user input  from `BOTH`, `LEADING`, `TRAINLING`  to the right int value (1,2,3), other input is invalid. so I think we have the precondition that the direction will only be 1,2,3. right ?

Reply:
Right! For this reason, I suggest to use box_err so that irregular or broken requests won't crash our server. It's a kind of possibility for "Denial of Service".